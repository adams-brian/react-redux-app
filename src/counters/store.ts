import { IAction } from '../store';

export interface IState {
  counters: number[];
}

/* tslint:disable:max-classes-per-file */
abstract class ActionWithNumberPayload implements IAction {
  public readonly type: string;
  constructor(public payload: number) {}
}
export const INCREMENT = '[Counters] INCREMENT';
class Increment extends ActionWithNumberPayload {
  public readonly type = INCREMENT;
}
export const DECREMENT = '[Counters] DECREMENT';
class Decrement extends ActionWithNumberPayload {
  public readonly type = DECREMENT;
}
export const RESET = '[Counters] RESET';
class Reset extends ActionWithNumberPayload {
  public readonly type = RESET;
}
export const ADD_COUNTER = '[Counters] ADD COUNTER';
class AddCounter implements IAction {
  public readonly type = ADD_COUNTER;
}
export const REMOVE_COUNTER = '[Counters] REMOVE COUNTER';
class RemoveCounter extends ActionWithNumberPayload {
  public readonly type = REMOVE_COUNTER;
}
export const COUNTERS_UPDATED = '[Counters] COUNTERS UPDATED';
class CountersUpdated implements IAction {
  public readonly type = COUNTERS_UPDATED;
  constructor(public payload: number[]) {}
}
export type CountersAction = Increment | Decrement | Reset | AddCounter | RemoveCounter | CountersUpdated;

export const actionCreators = {
  addCounter: () => Object.assign({}, new AddCounter()),
  countersUpdated: (counters: number[]) => Object.assign({}, new CountersUpdated(counters)),
  decrement: (index: number) => Object.assign({}, new Decrement(index)),
  increment: (index: number) => Object.assign({}, new Increment(index)),
  removeCounter: (index: number) => Object.assign({}, new RemoveCounter(index)),
  reset: (index: number) => Object.assign({}, new Reset(index)),
};

function countersReducer(state: number[] = [], action: CountersAction) {
  switch (action.type) {
    case INCREMENT:
      return [
        ...state.slice(0, action.payload),
        state[action.payload] + 1,
        ...state.slice(action.payload + 1)
      ];
    case DECREMENT:
      return [
        ...state.slice(0, action.payload),
        state[action.payload] - 1,
        ...state.slice(action.payload + 1)
      ];
    case RESET:
      return [
        ...state.slice(0, action.payload),
        0,
        ...state.slice(action.payload + 1)
      ];
    case ADD_COUNTER:
      return [...state, 0];
    case REMOVE_COUNTER:
      return [
        ...state.slice(0, action.payload),
        ...state.slice(action.payload + 1)
      ];
    case COUNTERS_UPDATED:
      return [...action.payload];
    default:
      return state;
  }
}

export const reducers = {
  counters: countersReducer
}
