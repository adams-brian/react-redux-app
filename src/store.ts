import { LocationChangeAction } from 'react-router-redux';

export interface IUser { 
  _id: string;
  firstname: string;
  lastname: string;
}

export interface IAppState {
  loading: boolean;
  counters: number[];
  users: IUser[];
}

interface IAction {
  type: string;
}

/* tslint:disable:max-classes-per-file */
export const START_LOADING = '[App] START LOADING';
class StartLoading implements IAction {
  public readonly type = START_LOADING;
}
export const DONE_LOADING = '[App] DONE LOADING';
class DoneLoading implements IAction {
  public readonly type = DONE_LOADING;
}
type LoadingAction = StartLoading | DoneLoading;

abstract class ActionWithNumberPayload implements IAction {
  public readonly type: string;
  constructor(public payload: number) {}
}
export const INCREMENT = '[Counters] INCREMENT';
class Increment extends ActionWithNumberPayload {
  public readonly type = INCREMENT;
}
export const DECREMENT = '[Counters] DECREMENT';
class Decrement extends ActionWithNumberPayload {
  public readonly type = DECREMENT;
}
export const RESET = '[Counters] RESET';
class Reset extends ActionWithNumberPayload {
  public readonly type = RESET;
}
export const ADD_COUNTER = '[Counters] ADD COUNTER';
class AddCounter implements IAction {
  public readonly type = ADD_COUNTER;
}
export const REMOVE_COUNTER = '[Counters] REMOVE COUNTER';
class RemoveCounter extends ActionWithNumberPayload {
  public readonly type = REMOVE_COUNTER;
}
export const COUNTERS_UPDATED = '[Counters] COUNTERS UPDATED';
class CountersUpdated implements IAction {
  public readonly type = COUNTERS_UPDATED;
  constructor(public payload: number[]) {}
}
type CounterAction = Increment | Decrement | Reset | AddCounter | RemoveCounter | CountersUpdated;

export const CREATE_USER = '[Users] CREATE';
export class CreateUser implements IAction {
  public readonly type = CREATE_USER;
  constructor(public payload: IUser) {}
}
export const UPDATE_USER = '[Users] UPDATE';
class UpdateUser implements IAction {
  public readonly type = UPDATE_USER;
  constructor(public payload: IUser) {}
}
export const DELETE_USER = '[Users] DELETE';
export class DeleteUser implements IAction {
  public readonly type = DELETE_USER;
  constructor(public payload: string) {}
}
export const USERS_UPDATED = '[Users] USERS UPDATED';
class UsersUpdated implements IAction {
  public readonly type = USERS_UPDATED;
  constructor(public payload: IUser[]) {}
}
type UserAction = CreateUser | UpdateUser | DeleteUser | UsersUpdated;

export const actionCreators = {
  doneLoading: () => Object.assign({}, new DoneLoading()),
  startLoading: () => Object.assign({}, new StartLoading()),

  addCounter: () => Object.assign({}, new AddCounter()),
  countersUpdated: (counters: number[]) => Object.assign({}, new CountersUpdated(counters)),
  decrement: (index: number) => Object.assign({}, new Decrement(index)),
  increment: (index: number) => Object.assign({}, new Increment(index)),
  removeCounter: (index: number) => Object.assign({}, new RemoveCounter(index)),
  reset: (index: number) => Object.assign({}, new Reset(index)),

  createUser: (user: IUser) => Object.assign({}, new CreateUser(user)),
  deleteUser: (id: string) => Object.assign({}, new DeleteUser(id)),
  updateUser: (user: IUser) => Object.assign({}, new UpdateUser(user)),
  usersUpdated: (users: IUser[]) => Object.assign({}, new UsersUpdated(users)),
};

function loadingReducer(state: boolean = false, action: LoadingAction) {
  switch (action.type) {
    case START_LOADING:
      return true;
    case DONE_LOADING:
      return false;
    default:
      return state;
  }
}

function countersReducer(state: number[] = [], action: CounterAction) {
  switch (action.type) {
    case INCREMENT:
      return [
        ...state.slice(0, action.payload),
        state[action.payload] + 1,
        ...state.slice(action.payload + 1)
      ];
    case DECREMENT:
      return [
        ...state.slice(0, action.payload),
        state[action.payload] - 1,
        ...state.slice(action.payload + 1)
      ];
    case RESET:
      return [
        ...state.slice(0, action.payload),
        0,
        ...state.slice(action.payload + 1)
      ];
    case ADD_COUNTER:
      return [...state, 0];
    case REMOVE_COUNTER:
      return [
        ...state.slice(0, action.payload),
        ...state.slice(action.payload + 1)
      ];
    case COUNTERS_UPDATED:
      return [...action.payload];
    default:
      return state;
  }
}

const usersReducer = (state: IUser[] = [], action: UserAction) => {
  switch (action.type) {
    case CREATE_USER:
      return [...state, action.payload];
    case UPDATE_USER:
      return state.map(user => user._id === action.payload._id ? Object.assign({}, action.payload) : user);
    case DELETE_USER:
      return state.filter(user => user._id !== action.payload);
    case USERS_UPDATED:
      return [...action.payload];
    default:
      return state;
  }
};

export const reducers = {
  counters: countersReducer,
  loading: loadingReducer,
  users: usersReducer
};

export type AppAction = LoadingAction | CounterAction | UserAction | LocationChangeAction;
