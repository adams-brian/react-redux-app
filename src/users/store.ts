import { Action, combineReducers } from 'redux';

export interface IUser {
  _id: string;
  firstname: string;
  lastname: string;
}

export interface IState {
  users: {
    error: string;
    list: IUser[];
    loaded: boolean;
  }
}

/* tslint:disable:max-classes-per-file */
export const CREATE_USER = '[Users] CREATE';
export class CreateUser implements Action {
  public readonly type = CREATE_USER;
  constructor(public payload: IUser) { }
}
export const UPDATE_USER = '[Users] UPDATE';
export class UpdateUser implements Action {
  public readonly type = UPDATE_USER;
  constructor(public payload: IUser) { }
}
export const DELETE_USER = '[Users] DELETE';
export class DeleteUser implements Action {
  public readonly type = DELETE_USER;
  constructor(public payload: string) { }
}
export const USERS_UPDATED = '[Users] USERS UPDATED';
class UsersUpdated implements Action {
  public readonly type = USERS_UPDATED;
  constructor(public payload: IUser[]) { }
}
export const LOAD_USERS = '[Users] LOAD USERS';
export class LoadUsers implements Action {
  public readonly type = LOAD_USERS;
}
export const USERS_ERROR = '[Users] USERS ERROR';
export class UsersError implements Action {
  public readonly type = USERS_ERROR;
  constructor(public payload: Error) { }
}
export const USERS_LOADED = '[Users] USERS LOADED';
export class UsersLoaded implements Action {
  public readonly type = USERS_LOADED;
}
export type UsersAction = CreateUser | UpdateUser | DeleteUser | UsersUpdated | LoadUsers | UsersError | UsersLoaded;

export const actionCreators = {
  createUser: (user: IUser) => Object.assign({}, new CreateUser(user)),
  deleteUser: (id: string) => Object.assign({}, new DeleteUser(id)),
  loadUsers: () => Object.assign({}, new LoadUsers()),
  updateUser: (user: IUser) => Object.assign({}, new UpdateUser(user)),
  usersError: (err: Error) => Object.assign({}, new UsersError(err)),
  usersLoaded: () => Object.assign({}, new UsersLoaded()),
  usersUpdated: (users: IUser[]) => Object.assign({}, new UsersUpdated(users)),
};

export const listReducer = (state: IUser[] = [], action: UsersAction) => {
  switch (action.type) {
    case CREATE_USER:
      return [...state, action.payload];
    case UPDATE_USER:
      return state.map(user => user._id === action.payload._id ? Object.assign({}, action.payload) : user);
    case DELETE_USER:
      return state.filter(user => user._id !== action.payload);
    case USERS_UPDATED:
      return [...action.payload];
    default:
      return state;
  }
};

export const loadedReducer = (state: boolean = false, action: UsersAction) => {
  switch (action.type) {
    case USERS_LOADED:
      return true;
    default:
      return state;
  }
}

export const errorReducer = (state: string = '', action: UsersAction) => {
  switch (action.type) {
    case USERS_ERROR:
      return action.payload.message;
    default:
      return state;
  }
}

export const reducers = {
  users: combineReducers({
    error: errorReducer,
    list: listReducer,
    loaded: loadedReducer
  })
};
