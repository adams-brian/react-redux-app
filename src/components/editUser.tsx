import * as React from 'react';
import { Link } from 'react-router-dom';

import { IUser } from '../store';

interface IEditUserProps extends IUser {
  submit: (user: IUser) => void;
}

export class EditUser extends React.Component<IEditUserProps, IUser> {
  constructor(props: IEditUserProps) {
    super(props);
    this.state = {
      _id: props._id,
      firstname: props.firstname,
      lastname: props.lastname
    };

    this.inputChanged = this.inputChanged.bind(this);
    this.submit = this.submit.bind(this);
  }

  public render() {
    return (
      <div className="edit-user">
        <h1>{this.state._id && this.state._id.length > 0 ? 'Edit' : 'Create'} User</h1>
        <form>
          <div className="form-group">
            <label htmlFor="firstname">Firstname</label>
            <input
              placeholder="Firstname"
              type="text"
              className="form-control"
              name="firstname"
              id="firstname"
              value={this.state.firstname}
              onChange={this.inputChanged}
              required={true}
            />
          </div>
          <div className="form-group">
            <label htmlFor="lastname">Lastname</label>
            <input
              placeholder="Lastname"
              type="text"
              className="form-control"
              name="lastname"
              id="lastname"
              value={this.state.lastname}
              onChange={this.inputChanged}
              required={true}
            />
          </div>
          <div>
            <Link to="/users">
              <button type="button" className="btn btn-secondary">Cancel</button>
            </Link>
            {' '}
            <button 
              className="btn btn-primary" 
              type="submit" 
              onClick={this.submit} 
              disabled={this.state.firstname.length === 0 || this.state.lastname.length === 0}
            >
              Submit
            </button>
          </div>
        </form>
      </div>
    );
  }

  private inputChanged = (e: React.FormEvent<HTMLInputElement>) => {
    this.setState({
      [e.currentTarget.name]: e.currentTarget.value
    } as Pick<IUser, keyof IUser>);
  }

  private submit(e: React.FormEvent<HTMLButtonElement>) {
    e.preventDefault();
    this.props.submit(this.state);
  }
}