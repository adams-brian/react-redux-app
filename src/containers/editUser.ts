import { connect, Dispatch } from 'react-redux';
import { AppState, actionCreators } from '../store';
import { RouteComponentProps } from 'react-router';
import { generate } from 'shortid';
import { push } from 'react-router-redux';

import { EditUser as UnwrappedEditUser } from '../components/editUser';
import { User, AppAction } from '../store';

export const EditUser = connect(
  (state: AppState, props: RouteComponentProps<{ id: string }>) => 
    state.users.find(u => u._id === props.match.params.id) || { _id: '', firstname: '', lastname: '' } as User,
  (dispatch: Dispatch<AppAction>) => ({
    submit: (u: User) => {
      if (u._id && u._id.length > 0) {
        dispatch(actionCreators.updateUser(u));
      }
      else {
        dispatch(actionCreators.createUser(Object.assign({}, u, { _id: generate() })));
      }
      dispatch(push('/users'));
    }
  })
  )(UnwrappedEditUser);
