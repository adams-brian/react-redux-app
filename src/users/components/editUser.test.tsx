import { shallow } from 'enzyme';
import { createLocation, createMemoryHistory } from 'history';
import * as React from 'react';
import { push } from 'react-router-redux';
import * as sinon from 'sinon';

import { actionCreators } from '../store';
import { EditUser, mapDispatchToProps, mapStateToProps } from './editUser';

describe("EditUser", () => {

  describe("component", () => {

    let element: JSX.Element;

    describe("create user mode", () => {

      const user = {
        _id: '',
        firstname: '',
        lastname: ''
      }
      let submit: sinon.SinonSpy;

      beforeEach(() => {
        submit = sinon.spy();
        element = (
          <EditUser
            user={user}
            submit={submit}
          />
        );
      });

      it('renders as expected', () => {
        const component = shallow(element);
        expect(component).toMatchSnapshot();
      });

      it('calls submit', () => {
        const component = shallow(element);
        expect(submit.called).toBe(false);
        component.find('#firstname').first().simulate('change', { currentTarget: { name: 'firstname', value: 'new' } });
        component.find('#lastname').first().simulate('change', { currentTarget: { name: 'lastname', value: 'user' } });
        component.find('button[type="submit"]').first().simulate('click', { preventDefault: () => { /* noop */ }});
        expect(submit.calledOnceWithExactly({ _id: '', firstname: 'new', lastname: 'user'})).toBe(true);
      });

    });

    describe("edit user mode", () => {

      const user = {
        _id: '123',
        firstname: 'abc',
        lastname: 'def'
      }
      let submit: sinon.SinonSpy;

      beforeEach(() => {
        submit = sinon.spy();
        element = (
          <EditUser
            user={user}
            submit={submit}
          />
        );
      });

      it('renders as expected', () => {
        const component = shallow(element);
        expect(component).toMatchSnapshot();
      });

      it('calls updateUser on the state', () => {
        const component = shallow(element);
        expect(submit.called).toBe(false);
        component.find('#firstname').first().simulate('change', { currentTarget: { name: 'firstname', value: 'updated' } });
        component.find('#lastname').first().simulate('change', { currentTarget: { name: 'lastname', value: 'user' } });
        component.find('button[type="submit"]').first().simulate('click', { preventDefault: () => { /* noop */ }});
        expect(submit.calledOnceWithExactly({ _id: '123', firstname: 'updated', lastname: 'user'})).toBe(true);
      });
      
    });

  });

  describe("connection", () => {

    const users = [
      {
        _id: '123',
        firstname: 'abc',
        lastname: 'def'
      },
      {
        _id: '456',
        firstname: 'hij',
        lastname: 'klm'
      },
      {
        _id: '789',
        firstname: 'nop',
        lastname: 'qrs'
      }
    ];

    it('should map state to props', () => {
      const getProps = (id: string) => 
        ({
          history: createMemoryHistory(),
          location: createLocation('/a/location'),
          match: {
            isExact: true,
            params: { id },
            path: '/a/location',
            url: 'urlbase/a/location'
          }
        });
      expect(mapStateToProps({ users }, getProps(''))).toEqual({
        user: {
          _id: '',
          firstname: '',
          lastname: ''
        }
      });
      expect(mapStateToProps({ users }, getProps('456'))).toEqual({
        user: {
          _id: '456',
          firstname: 'hij',
          lastname: 'klm'
        }
      });
    });

    it('dispatches submit for new user', () => {
      const dispatch = sinon.spy();
      const props = mapDispatchToProps(dispatch);
      expect(dispatch.called).toBe(false);
      const user = {
        _id: '',
        firstname: 'new',
        lastname: 'user'
      }
      props.submit(user);
      expect(dispatch.calledTwice).toBe(true);
      // can't use property matchers in typescript
      expect(dispatch.firstCall.args.length).toBe(1);
      const arg = dispatch.firstCall.args[0];
      expect(arg.payload._id.length > 0).toBe(true);
      delete arg.payload._id;
      expect(arg).toMatchSnapshot();
      expect(dispatch.secondCall.calledWithExactly(push('/users'))).toBe(true);
    });

    it('dispatches submit for an updated user', () => {
      const dispatch = sinon.spy();
      const props = mapDispatchToProps(dispatch);
      expect(dispatch.called).toBe(false);
      const user = {
        _id: '123',
        firstname: 'abc',
        lastname: 'def'
      }
      props.submit(user);
      expect(dispatch.calledTwice).toBe(true);
      expect(dispatch.firstCall.calledWithExactly(actionCreators.updateUser(user))).toBe(true);
      expect(dispatch.secondCall.calledWithExactly(push('/users'))).toBe(true);
    });

  });

});
