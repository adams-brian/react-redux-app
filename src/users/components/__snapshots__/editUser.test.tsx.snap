// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditUser component create user mode renders as expected 1`] = `
<div
  className="edit-user"
>
  <h1>
    Create
     User
  </h1>
  <form>
    <div
      className="form-group"
    >
      <label
        htmlFor="firstname"
      >
        Firstname
      </label>
      <input
        className="form-control"
        id="firstname"
        name="firstname"
        onChange={[Function]}
        placeholder="Firstname"
        required={true}
        type="text"
        value=""
      />
    </div>
    <div
      className="form-group"
    >
      <label
        htmlFor="lastname"
      >
        Lastname
      </label>
      <input
        className="form-control"
        id="lastname"
        name="lastname"
        onChange={[Function]}
        placeholder="Lastname"
        required={true}
        type="text"
        value=""
      />
    </div>
    <div>
      <Link
        replace={false}
        to="/users"
      >
        <button
          className="btn btn-secondary"
          type="button"
        >
          Cancel
        </button>
      </Link>
       
      <button
        className="btn btn-primary"
        disabled={true}
        onClick={[Function]}
        type="submit"
      >
        Submit
      </button>
    </div>
  </form>
</div>
`;

exports[`EditUser component edit user mode renders as expected 1`] = `
<div
  className="edit-user"
>
  <h1>
    Edit
     User
  </h1>
  <form>
    <div
      className="form-group"
    >
      <label
        htmlFor="firstname"
      >
        Firstname
      </label>
      <input
        className="form-control"
        id="firstname"
        name="firstname"
        onChange={[Function]}
        placeholder="Firstname"
        required={true}
        type="text"
        value="abc"
      />
    </div>
    <div
      className="form-group"
    >
      <label
        htmlFor="lastname"
      >
        Lastname
      </label>
      <input
        className="form-control"
        id="lastname"
        name="lastname"
        onChange={[Function]}
        placeholder="Lastname"
        required={true}
        type="text"
        value="def"
      />
    </div>
    <div>
      <Link
        replace={false}
        to="/users"
      >
        <button
          className="btn btn-secondary"
          type="button"
        >
          Cancel
        </button>
      </Link>
       
      <button
        className="btn btn-primary"
        disabled={false}
        onClick={[Function]}
        type="submit"
      >
        Submit
      </button>
    </div>
  </form>
</div>
`;

exports[`EditUser connection dispatches submit for new user 1`] = `
Object {
  "payload": Object {
    "firstname": "new",
    "lastname": "user",
  },
  "type": "[Users] CREATE",
}
`;
